File: src/ios/TalsecPlugin.swift
import Foundation
import TalsecRuntime

@objc(TalsecPlugin) class TalsecPlugin : CDVPlugin {
    public static var shared:TalsecPlugin?
    
    override func pluginInitialize() {
        TalsecContext.context.commandDelegate = self.commandDelegate
        TalsecPlugin.shared = self
    }
    
    @objc(getThreatChannelData:)
    private func getThreatChannelData(command: CDVInvokedUrlCommand) -> Void {
        TalsecContext.sendMessage(msg: [EventIdentifiers.threatChannelKey], callbackId: command.callbackId, keepCallback: false)
    }
    
    /**
     * Method to get the random identifiers of callbacks
     */
    @objc(getThreatIdentifiers:)
    private func getThreatIdentifiers(command: CDVInvokedUrlCommand) -> Void {
        let threatIdentifiers = getCordovaThreatIdentifiers()
        TalsecContext.sendMessage(msg: threatIdentifiers, callbackId: command.callbackId, keepCallback: false)
    }
    
    @objc(getRaspExecutionStateChannelData:)
    private func getRaspExecutionStateChannelData(command: CDVInvokedUrlCommand) -> Void {
        TalsecContext.sendMessage(msg: [EventIdentifiers.raspExecutionStateChannelKey], callbackId: command.callbackId, keepCallback: false)
    }
    
    /**
     * Method to get the random identifiers of callbacks
     */
    @objc(getRaspExecutionStateIdentifiers:)
    private func getRaspExecutionStateIdentifiers(command: CDVInvokedUrlCommand) -> Void {
        let threatIdentifiers = getCordovaRaspExecutionStateIdentifiers()
        TalsecContext.sendMessage(msg: threatIdentifiers, callbackId: command.callbackId, keepCallback: false)
    }
    
    @objc(start:)
    func start(command: CDVInvokedUrlCommand) {
        guard let talsecConfig = command.arguments[0] as? NSDictionary else {
            TalsecContext.sendError(msg: "Missing config parameter in Talsec Native Plugin", callbackId: command.callbackId)
            return
        }
        do {
            try initializeTalsec(talsecConfig: talsecConfig)
        }
        catch let error as NSError {
            TalsecContext.sendError(msg: "Could not initialize freeRASP: \(error.domain)", callbackId: command.callbackId)
            return
        }
        TalsecContext.sendMessage(msg: "started", callbackId: command.callbackId, keepCallback: false)
    }
  
    @objc(registerListener:)
    private func registerListener(command: CDVInvokedUrlCommand) {
        TalsecContext.context.threatCallbackCordova = command.callbackId
        ThreatDispatcher.shared.listener = { threat in
            TalsecContext.sendMessage(msg: [EventIdentifiers.threatChannelKey : threat.callbackIdentifier], callbackId: command.callbackId)
        }
    }

    @objc(registerRaspExecutionStateListener:)
    private func registerRaspExecutionStateListener(command: CDVInvokedUrlCommand) {
        TalsecContext.context.raspExecutionStateCallbackCordova = command.callbackId
        ExecutionStateDispatcher.shared.listener = { event in
             TalsecContext.sendMessage(msg: [EventIdentifiers.raspExecutionStateChannelKey : event.callbackIdentifier], callbackId: command.callbackId)
        }
    }
    
    @objc(onInvalidCallback:)
    private func onInvalidCallback(command: CDVInvokedUrlCommand) -> Void {
        abort()
    }

    @objc(blockScreenCapture:)
    private func blockScreenCapture(command: CDVInvokedUrlCommand) -> Void {
        guard let enable = (command.arguments[0] as? Bool) else {
            TalsecContext.sendError(msg: "Wrong format of the enable argument", callbackId: command.callbackId)
            return
        }
        
        getProtectedWindow { window in
            if let window = window {
                Talsec.blockScreenCapture(enable: enable, window: window)
                TalsecContext.sendMessage(msg: "OK", callbackId: command.callbackId)
            } else {
                TalsecContext.sendError(msg: "No windows found to block screen capture", callbackId: command.callbackId)
            }
        }
    }
    
    @objc(isScreenCaptureBlocked:)
    private func isScreenCaptureBlocked(command: CDVInvokedUrlCommand) -> Void {
        getProtectedWindow { window in
            if let window = window {
                let isBlocked = Talsec.isScreenCaptureBlocked(in: window)
                TalsecContext.sendMessage(msg: isBlocked ? ScreenCaptureStatus.blocked.rawValue : ScreenCaptureStatus.allowed.rawValue, callbackId: command.callbackId)
            } else {
                TalsecContext.sendError(msg: "Error while checking if screen capture is blocked", callbackId: command.callbackId)
            }
        }
    }
    
    private func getProtectedWindow(completion: @escaping (UIWindow?) -> Void) {
        DispatchQueue.main.async {
            if #available(iOS 13.0, *) {
                if let windowScene = UIApplication.shared.connectedScenes.first as? UIWindowScene {
                    if let window = windowScene.windows.first {
                        completion(window)
                    } else {
                        completion(nil)
                    }
                } else {
                    completion(nil)
                }
            }
        }
    }

    @objc(storeExternalId:)
    private func storeExternalId(command: CDVInvokedUrlCommand) {
        guard let data = command.arguments[0] as? String else {
            TalsecContext.sendError(msg: "External id must be String", callbackId: command.callbackId)
            return
        }
        UserDefaults.standard.set(data, forKey: "app.talsec.externalid")
        TalsecContext.sendMessage(msg: "OK", callbackId: command.callbackId, keepCallback: false)
    }
    
    @objc(removeExternalId:)
    private func removeExternalId(command: CDVInvokedUrlCommand) {
        UserDefaults.standard.removeObject(forKey: "app.talsec.externalid")
        TalsecContext.sendMessage(msg: "OK", callbackId: command.callbackId, keepCallback: false)
    }
    
    func initializeTalsec(talsecConfig: NSDictionary) throws {
        guard let iosConfig = talsecConfig["iosConfig"] as? NSDictionary else {
            throw NSError(domain: "Missing iosConfig parameter in Talsec Native Plugin", code: 1)
        }
        guard let appBundleIds = iosConfig["appBundleIds"] as? String else {
            throw NSError(domain: "Missing appBundleIds parameter in Talsec Native Plugin", code: 2)
        }
        guard let appTeamId = iosConfig["appTeamId"] as? String else {
            throw NSError(domain: "Missing appTeamId parameter in Talsec Native Plugin", code: 3)
        }
        guard let watcherMailAddress = talsecConfig["watcherMail"] as? String else {
            throw NSError(domain: "Missing watcherMail parameter in Talsec Native Plugin", code: 4)
        }
        let isProd = talsecConfig["isProd"] as? Bool ?? true
        
        let config = TalsecConfig(appBundleIds: [appBundleIds], appTeamId: appTeamId, watcherMailAddress: watcherMailAddress, isProd: isProd)
        commandDelegate.run(inBackground: {
            Talsec.start(config: config)
        })
    }
}

extension SecurityThreatCenter: @retroactive SecurityThreatHandler, @retroactive RaspExecutionState {
    
    public func threatDetected(_ securityThreat: TalsecRuntime.SecurityThreat) {
        if (securityThreat.rawValue == "passcodeChange") {
            return
        }
        ThreatDispatcher.shared.dispatch(threat: securityThreat)
    }
    
    public func onAllChecksFinished() {
        ExecutionStateDispatcher.shared.dispatch(event: RaspExecutionStates.allChecksFinished)
    }
}

class TalsecContext : CDVPlugin {
    static let context = TalsecContext()
    var threatCallbackCordova: String?
    var raspExecutionStateCallbackCordova: String?

    static func sendMessage(msg: Any, callbackId: String, keepCallback: Bool = true) {
        // send the result to JavaScript on the main thread
        DispatchQueue.main.async {
            var pluginResult: CDVPluginResult?

            if let intMsg = msg as? Int {
                pluginResult = CDVPluginResult(
                    status: CDVCommandStatus.ok,
                    messageAs: intMsg
                )
            } else if let stringMsg = msg as? String {
                pluginResult = CDVPluginResult(
                    status: CDVCommandStatus.ok,
                    messageAs: stringMsg
                )
            } else if let arrayMsg = msg as? [Any] {
                pluginResult = CDVPluginResult(
                    status: CDVCommandStatus.ok,
                    messageAs: arrayMsg
                )
            } else if let dictMsg = msg as? [AnyHashable: Any] {
                pluginResult = CDVPluginResult(
                    status: CDVCommandStatus.ok,
                    messageAs: dictMsg
                )
            }

            if let result = pluginResult {
                result.setKeepCallbackAs(keepCallback)
                
                context.commandDelegate?.send(
                    result,
                    callbackId: callbackId
                )
            } else {
                TalsecContext.sendError(msg: "Unsupported operation", callbackId: callbackId)
            }
        }
    }

    static func sendError(msg: String, callbackId: String) {
        DispatchQueue.main.async {
            // send the result to JavaScript on the main thread
            let pluginResult = CDVPluginResult(
                status: CDVCommandStatus.error,
                messageAs: msg
            )
            
            context.commandDelegate?.send(
                pluginResult,
                callbackId: callbackId
            )
        }
    }
}

File: src/android/TalsecPlugin.kt
package com.aheaditec.talsec.cordova

import android.content.Context
import android.os.Build
import android.os.Handler
import android.os.HandlerThread
import android.os.Looper
import android.util.Log
import com.aheaditec.talsec.cordova.events.BaseRaspEvent
import com.aheaditec.talsec.cordova.events.RaspExecutionStateEvent
import com.aheaditec.talsec.cordova.events.ThreatEvent
import com.aheaditec.talsec.cordova.interfaces.PluginExecutionStateListener
import com.aheaditec.talsec.cordova.interfaces.PluginThreatListener
import com.aheaditec.talsec.cordova.utils.Utils
import com.aheaditec.talsec.cordova.utils.ScreenCaptureStatus
import com.aheaditec.talsec.cordova.utils.toEncodedJsonArray
import com.aheaditec.talsec.cordova.utils.getArraySafe
import com.aheaditec.talsec.cordova.utils.getBooleanSafe
import com.aheaditec.talsec.cordova.utils.getNestedArraySafe
import com.aheaditec.talsec.cordova.utils.getStringSafe
import com.aheaditec.talsec_security.security.api.SuspiciousAppInfo
import com.aheaditec.talsec_security.security.api.Talsec
import com.aheaditec.talsec_security.security.api.TalsecConfig

import org.apache.cordova.CallbackContext
import org.apache.cordova.CordovaPlugin
import org.apache.cordova.PluginResult
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

class TalsecPlugin : CordovaPlugin() {

    override fun execute(
        action: String?, args: JSONArray?, callbackContext: CallbackContext?
    ): Boolean {
        return when (action) {
            "start" -> start(args, callbackContext)
            "onInvalidCallback" -> onInvalidCallback(callbackContext)
            "getThreatIdentifiers" -> getThreatIdentifiers(callbackContext)
            "getRaspExecutionStateIdentifiers" -> getRaspExecutionStateIdentifiers(callbackContext)
            "getThreatChannelData" -> getThreatChannelData(callbackContext)
            "getRaspExecutionStateChannelData" -> getRaspExecutionStateChannelData(callbackContext)
            "addToWhitelist" -> addToWhitelist(callbackContext, args)
            "getAppIcon" -> getAppIcon(callbackContext, args)
            "blockScreenCapture" -> blockScreenCapture(callbackContext, args)
            "isScreenCaptureBlocked" -> isScreenCaptureBlocked(callbackContext)
            "storeExternalId" -> storeExternalId(callbackContext, args)
            "removeExternalId" -> removeExternalId(callbackContext)
            "registerListener" -> registerListener(callbackContext)
            "registerRaspExecutionStateListener" -> registerRaspExecutionStateListener(
                callbackContext
            )

            else -> {
                callbackContext?.error("Talsec plugin executed with unknown action - $action")
                return false
            }
        }
    }

    override fun pluginInitialize() {
        super.pluginInitialize()
        // Trigger lazy initialization of the events
        ThreatEvent.ALL_EVENTS
        RaspExecutionStateEvent.ALL_EVENTS
        TalsecThreatHandler.initializeDispatchers(PluginListener(this.cordova.context))
    }

    override fun onStart() {
        super.onStart()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
            ScreenProtector.register(cordova.activity)
        }
    }

    override fun onPause(multitasking: Boolean) {
        super.onPause(multitasking)
        TalsecThreatHandler.threatDispatcher.onPause()
        TalsecThreatHandler.executionStateDispatcher.onPause()
    }

    override fun onStop() {
        super.onStop()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
            ScreenProtector.unregister(cordova.activity)
        }
        TalsecThreatHandler.threatDispatcher.unregisterListener()
        TalsecThreatHandler.executionStateDispatcher.unregisterListener()
    }

    override fun onResume(multitasking: Boolean) {
        super.onResume(multitasking)
        TalsecThreatHandler.threatDispatcher.onResume()
        TalsecThreatHandler.executionStateDispatcher.onResume()
    }

    override fun onDestroy() {
        super.onDestroy()
        backgroundHandlerThread.quitSafely()
    }

    private fun start(args: JSONArray?, callbackContext: CallbackContext?): Boolean {
        if (talsecStarted) {
            callbackContext?.success("started")
            return true
        }

        val configJson = args?.optString(0, null) ?: kotlin.run {
            callbackContext?.error("Missing config parameter in Talsec Native Plugin")
            return false
        }

        try {
            val config = buildTalsecConfigThrowing(configJson)
            TalsecThreatHandler.registerListener(this.cordova.context)
            this.cordova.activity.runOnUiThread {
                Talsec.start(this.cordova.context, config)
                mainHandler.post {
                    talsecStarted = true
                    // This code must be called only AFTER Talsec.start
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.UPSIDE_DOWN_CAKE) {
                        ScreenProtector.register(cordova.activity)
                    }
                }
            }
            callbackContext?.success("started")
            return true
        } catch (e: Exception) {
            callbackContext?.error("Error during Talsec Native plugin initialization - ${e.message}")
        }
        return false
    }

    private fun addToWhitelist(callbackContext: CallbackContext?, args: JSONArray?): Boolean {
        val packageName = args?.optString(0, null) ?: run {
            callbackContext?.error("Missing packageName parameter in Talsec Native Plugin")
            return false
        }
        this.cordova.activity.runOnUiThread {
            Talsec.addToWhitelist(cordova.context, packageName)
        }
        callbackContext?.success("OK - package $packageName added to whitelist")
        return true
    }

    /**
     * Method retrieves app icon for the given parameter
     * @param args - should contain `packageName`: package name of the app we want to retrieve icon for
     * @return PNG with app icon encoded as a base64 string
     */
    private fun getAppIcon(callbackContext: CallbackContext?, args: JSONArray?): Boolean {
        val packageName = args?.optString(0, null) ?: run {
            callbackContext?.error("Missing packageName parameter in Talsec Native Plugin")
            return false
        }
        // Perform the app icon encoding on a background thread
        backgroundHandler.post {
            val encodedData = Utils.getAppIconAsBase64String(cordova.context, packageName)
            mainHandler.post { callbackContext?.success(encodedData) }
        }
        return true
    }

    private fun blockScreenCapture(callbackContext: CallbackContext?, args: JSONArray?): Boolean {
        try {
            val enable = args?.getBoolean(0) ?: run {
                callbackContext?.error("Missing enable parameter in Talsec Native Plugin")
                return false
            }
            cordova.activity.runOnUiThread {
                try {
                    Talsec.blockScreenCapture(cordova.activity, enable)
                    callbackContext?.success("Screen capture is now ${if (enable) "Blocked" else "Enabled"}.")
                } catch (e: Exception) {
                    callbackContext?.error("Failed to block screen capture: ${e.message}")
                }
            }
            return true
        } catch (e: JSONException) {
            callbackContext?.error("Invalid argument: Expected a boolean value")
            return false
        } catch (e: Exception) {
            callbackContext?.error("Failed to block screen capture: ${e.message}")
            return false
        }
    }

    private fun isScreenCaptureBlocked(callbackContext: CallbackContext?): Boolean {
        try {
            val isBlocked = Talsec.isScreenCaptureBlocked()
            val status = if (isBlocked) ScreenCaptureStatus.BLOCKED else ScreenCaptureStatus.ALLOWED
            callbackContext?.success(status)
            return true
        } catch (e: Exception) {
            callbackContext?.error("Failed to check screen capture status: ${e.message}")
            return false
        }
    }

    private fun storeExternalId(
        callbackContext: CallbackContext?, args: JSONArray?
    ): Boolean {
        val externalIdValue = args?.optString(0, null) ?: kotlin.run {
            callbackContext?.error("Missing external ID parameter in freeRASP Native Plugin")
            return false
        }

        return try {
            Talsec.storeExternalId(cordova.context, externalIdValue)
            callbackContext?.success("OK.")
            true
        } catch (e: Exception) {
            callbackContext?.error("Error during storeExternalId operation in freeRASP Native Plugin")
            false
        }
    }

    private fun removeExternalId(callbackContext: CallbackContext?): Boolean {
        return try {
            Talsec.removeExternalId(cordova.context)
            callbackContext?.success("OK.")
            true
        } catch (e: Exception) {
            callbackContext?.error("Error during removeExternalId operation in freeRASP Native Plugin")
            false
        }
    }

    /**
     * We never send an invalid callback over our channel.
     * Therefore, if this happens, we want to kill the app.
     */
    private fun onInvalidCallback(_callbackContext: CallbackContext?): Boolean {
        android.os.Process.killProcess(android.os.Process.myPid())
        return true
    }

    /**
     * Method to get the random identifiers of callbacks
     */
    private fun getThreatIdentifiers(callbackContext: CallbackContext?): Boolean {
        callbackContext?.success(ThreatEvent.ALL_EVENTS)
        return true
    }

    /**
     * Method to get the random identifiers of callbacks
     */
    private fun getRaspExecutionStateIdentifiers(callbackContext: CallbackContext?): Boolean {
        callbackContext?.success(RaspExecutionStateEvent.ALL_EVENTS)
        return true
    }

    /**
     * Method to setup the message passing between native and Cordova
     * @return list of [ThreatEvent.CHANNEL_KEY, ThreatEvent.MALWARE_CHANNEL_KEY]
     */
    private fun getThreatChannelData(callbackContext: CallbackContext?): Boolean {
        val channelData = JSONArray()
        channelData.put(ThreatEvent.CHANNEL_KEY)
        channelData.put(ThreatEvent.MALWARE_CHANNEL_KEY)
        callbackContext?.success(channelData)
        return true
    }

    /**
     * Method to setup the message passing between native and Cordova
     * @return list of [RaspExecutionStateEvent.CHANNEL_KEY]
     */
    private fun getRaspExecutionStateChannelData(callbackContext: CallbackContext?): Boolean {
        val channelData = JSONArray()
        channelData.put(RaspExecutionStateEvent.CHANNEL_KEY)
        callbackContext?.success(channelData)
        return true
    }

    private fun registerListener(callbackContext: CallbackContext?): Boolean {
        TalsecThreatHandler.threatDispatcher.registerListener(callbackContext)
        return true
    }

    private fun registerRaspExecutionStateListener(callbackContext: CallbackContext?): Boolean {
        TalsecThreatHandler.executionStateDispatcher.registerListener(callbackContext)
        return true
    }

    private fun buildTalsecConfigThrowing(configJson: String): TalsecConfig {
        val json = JSONObject(configJson)
        val androidConfig = json.getJSONObject("androidConfig")
        val packageName = androidConfig.getString("packageName")
        val certificateHashes = androidConfig.getArraySafe("certificateHashes")
        val talsecBuilder = TalsecConfig.Builder(packageName, certificateHashes)
            .watcherMail(json.getStringSafe("watcherMail"))
            .supportedAlternativeStores(androidConfig.getArraySafe("supportedAlternativeStores"))
            .prod(json.getBooleanSafe("isProd"))
            .killOnBypass(json.getBooleanSafe("killOnBypass", false))

        if (androidConfig.has("malwareConfig")) {
            val malwareConfig = androidConfig.getJSONObject("malwareConfig")
            talsecBuilder.whitelistedInstallationSources(malwareConfig.getArraySafe("whitelistedInstallationSources"))
            talsecBuilder.blacklistedHashes(malwareConfig.getArraySafe("blacklistedHashes"))
            talsecBuilder.blacklistedPackageNames(malwareConfig.getArraySafe("blacklistedPackageNames"))
            talsecBuilder.suspiciousPermissions(malwareConfig.getNestedArraySafe("suspiciousPermissions"))
        }
        return talsecBuilder.build()
    }

    companion object {
        private val backgroundHandlerThread = HandlerThread("BackgroundThread").apply { start() }
        private val backgroundHandler = Handler(backgroundHandlerThread.looper)
        private val mainHandler = Handler(Looper.getMainLooper())

        internal var talsecStarted = false
    }

    internal class PluginListener(
        private val context: Context
    ) : PluginThreatListener, PluginExecutionStateListener {

        override var threatCallbackContext: CallbackContext? = null
        override var executionStateCallbackContext: CallbackContext? = null

        override fun threatDetected(threatEventType: ThreatEvent) {
            val params = JSONObject()
            params.put(threatEventType.channelKey, threatEventType.value)
            sendResult(params, threatCallbackContext)
        }

        override fun malwareDetected(suspiciousApps: MutableList<SuspiciousAppInfo>) {
            backgroundHandler.post {
                val encodedSuspiciousApps = suspiciousApps.toEncodedJsonArray(context)
                mainHandler.post {
                    val params = JSONObject()
                    params.put(ThreatEvent.CHANNEL_KEY, ThreatEvent.Malware.value)
                    params.put(ThreatEvent.MALWARE_CHANNEL_KEY, encodedSuspiciousApps)
                    sendResult(params, threatCallbackContext)
                }
            }
        }

        override fun raspExecutionStateChanged(event: RaspExecutionStateEvent) {
            val params = JSONObject()
            params.put(event.channelKey, event.value)
            sendResult(params, executionStateCallbackContext)
        }

        private fun sendResult(json: JSONObject, callbackContext: CallbackContext?) {
            if (callbackContext != null) {
                val result = PluginResult(PluginResult.Status.OK, json)
                result.keepCallback = true
                callbackContext.sendPluginResult(result)
            } else {
                Log.w("TalsecPlugin", "Listener not registered.")
            }
        }
    }
}
